denops/@ddc-sources/path.ts

```diff
@@ -19,33 +19,68 @@ type Params = {
   cmd: string[];
   absolute: boolean;
   dirSeparator: DirSeparator | "";
+  escapeChars: string;
 };
+type WordFilter = (word: string) => string;
 
-function isValidUnixPath(path: string): boolean {
-  let escape = false;
-  for (let i = 0; i < path.length; i++) {
-    if (!escape && path[i].match(/\s/)) {
+function isValidUnixPath(path: string, escapeChars: string): boolean {
+  for (let i = 0; i < path.length; ++i) {
+    if (
+      path[i] === "\\" && (i + 1) < path.length &&
+      escapeChars.includes(path[i + 1])
+    ) {
+      ++i;
+    } else if (escapeChars.includes(path[i])) {
       return false;
     }
-    if (path[i] === "\\") {
-      escape = !escape;
-    } else {
-      escape = false;
+  }
+  return true;
+}
+
+function isValidWindowsPath(path: string, escapeChars: string): boolean {
+  // allow both '/' and '\'
+  const invalidChars = ':*?"<>|' + escapeChars;
+  let i = 0;
+  if (/^[a-zA-Z]:/.test(path)) {
+    i = 2;
+  }
+  for (; i < path.length; ++i) {
+    if (
+      path[i] === "\\" && (i + 1) < path.length &&
+      escapeChars.includes(path[i + 1])
+    ) {
+      ++i;
+    } else if (invalidChars.includes(path[i])) {
+      return false;
     }
   }
   return true;
 }
 
+function getWordFilter(escapeChars: string): WordFilter {
+  if (escapeChars.length === 0) {
+    return (word) => word;
+  }
+  const escapeRegex = new RegExp(
+    `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`,
+    "g",
+  );
+  return (word) => word.replaceAll(escapeRegex, "\\$&");
+}
```

Comment:
Consider using `escape()` defined as  vim builtin function.

Reply:
`getWordFilter()` called many times.
Implemented in Deno to offload communication overhead of denops and processing in Vim.

Reply:
Noted. Then, How about the following code?

```js
function getWordFilter(escapeChars: string): WordFilter {
  const regexpsrc = `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`
  const escapeRegex = new RegExp(regexpsrc, "g");
  return (word) => word.replaceAll(escapeRegex, "\\$&");
}
```

We know that the Deno/ V8 is **very fast** interpreter. The creation of regexp objects is not so matter for the performance, is it? Further, we can expect cashing the regexp objects by the optimization of V8. Thus, the closure is no Moreover, '/[]/' matches nothing. I pefer the simpler code. Shall we believe the V8 optimization?

Reply:
Can you change the code block to a suggestion?

Reply:
Suggestion merged!